File: frontend/src/lib/factorTypes.js
// 策略 → 因子类型 映射（前端展示副本）
//
// ⚠️ 单一真源在后端：smcore/strategy/factor_types.py。
// 改这里前请先改后端并同步——两处必须一致，否则报告/监控层与前端分类会漂移。
// 一致性由 tests/test_frontend_registry_parity.py 守卫（键集合 + 类型串 + 顺序 +
// 配色覆盖逐一比对），漂移会直接红灯，不再靠人眼发现。
//
// 设计原则（与后端一致，用户 2026-09-15 要求「按因子类型分类」）：
// - 每策略归一个主因子类型（1:1），保证归并桶清晰、无歧义；
// - 一个标的可命中多策略 → 可映射到多个因子类型；
// - 因子类型顺序 / 配色与既有策略认知一致，便于阅读。
//
// 2026-09-17 重构：用户要求「完全去掉之前的策略，重新整一套因子候选」。原 boll/relativity/
// theme/cctv/momentum 复合策略、基本面 quality/value/size 复合体、boll/relativity 的 6 个原子、
// C 批基本面单指标原子（ROE/毛利率/EP/BP）与反向动量 lowmom20 全部移除，菜单收缩为
// **A 批 12 个因子池（factor_zoo 预注册文法 v1）存活价格因子**：
//   量价相关(pvcorr20/60) / 成交稳定(cvamt20/60) / 收益偏度(skew20/60) /
//   波动比(vratio20_120/vratio10_60) / 位置·距低点(distlo10/60) / 波动(vol20) / 非流动性(illiq20)。
// 这些因子公式 1:1 复用 factor_zoo.compute_factor，已验证存活。新因子候选（来自因子挖掘系统 /
// 开源 Alpha101 等）后续按同一注册表协议接入：在 factor_types 注册 + 产出同名
// Stock-Selection-<Label>-<date>.csv 即可，前端三处副本（本文件 / styles.css / App.jsx）同步更新，
// tests/test_frontend_registry_parity.py 守卫一致性。
//
// 配色约定：**同族用同色系不同明度**（如 pvcorr20/60 同属品红、cvamt20/60 同属紫、skew20/60 同属蓝），
// 便于一眼区分窗口变体。

export const STRATEGY_FACTOR_TYPE = {
  // A 批：因子池存活价格因子（菜单当前全集）
  pvcorr20: '量价相关',
  pvcorr60: '量价相关',
  cvamt20: '成交稳定',
  cvamt60: '成交稳定',
  skew20: '收益偏度',
  skew60: '收益偏度',
  vratio20_120: '波动比',
  vratio10_60: '波动比',
  distlo10: '位置·距低点',
  distlo60: '位置·距低点',
  vol20: '波动',
  illiq20: '非流动性',
}

export const FACTOR_TYPE_ORDER = [
  '量价相关',
  '成交稳定',
  '收益偏度',
  '波动比',
  '位置·距低点',
  '波动',
  '非流动性',
  '其他',
]

const _DEFAULT_TYPE = '其他'

function _normalize(name) {
  return String(name ?? '').trim().toLowerCase()
}

// 来源策略串（如 'Boll/Momentum' / 'cctv' / 'Boll、Momentum'）→ 去重后的因子类型列表。
// 用于明细展示与贡献度统计：一个 pick 可能命中多个策略，故可能返回多个类型。
export function factorTypesOfSource(source) {
  if (!source) return [_DEFAULT_TYPE]
  const out = []
  for (const part of String(source).replace(/[、,]/g, '/').split('/')) {
    const k = _normalize(part)
    if (!k) continue
    const t = STRATEGY_FACTOR_TYPE[k] || _DEFAULT_TYPE
    if (!out.includes(t)) out.push(t)
  }
  return out.length ? out : [_DEFAULT_TYPE]
}

// 因子「家族」= 类型串中「·」之前的部分（如 '相对强度·抗跌满足率' → '相对强度'）。
// 拆分后同一族会有多个类型串，调用方**不要**再按完整串精确匹配。
export function typeFamilyOf(type) {
  const t = String(type ?? '')
  const i = t.indexOf('·')
  return i >= 0 ? t.slice(0, i) : t
}

// 来源策略串是否命中某家族（'相对强度' / '反转' / '基本面' …）。
// 后端 fusion 趋势闸门用的是同一套前缀归并，前后端口径保持一致。
export function hasTypeFamily(source, family) {
  const f = String(family ?? '')
  if (!f) return false
  return factorTypesOfSource(source).some((t) => t === f || t.startsWith(`${f}·`))
}

// 因子类型配色（与策略配色风格一致，用于前端彩色标签 / 左边框 / 分布条）
export const FACTOR_TYPE_COLORS = {
  // A 批：量价 / 微观结构族（菜单当前全集；与后端 factor_types.FACTOR_TYPE_ORDER 一致）
  '量价相关': { bg: 'hsla(330, 81%, 45%, 0.10)', text: '#DB2777', border: 'hsla(330, 81%, 45%, 0.35)' },
  '成交稳定': { bg: 'hsla(292, 84%, 41%, 0.10)', text: '#A21CAF', border: 'hsla(292, 84%, 41%, 0.35)' },
  '收益偏度': { bg: 'hsla(224, 76%, 40%, 0.10)', text: '#1E40AF', border: 'hsla(224, 76%, 40%, 0.35)' },
  '波动比': { bg: 'hsla(189, 94%, 35%, 0.10)', text: '#0891B2', border: 'hsla(189, 94%, 35%, 0.35)' },
  '位置·距低点': { bg: 'hsla(30, 6%, 32%, 0.10)', text: '#57534E', border: 'hsla(30, 6%, 32%, 0.35)' },
  '波动': { bg: 'hsla(43, 96%, 40%, 0.10)', text: '#CA8A04', border: 'hsla(43, 96%, 40%, 0.35)' },
  '非流动性': { bg: 'hsla(80, 79%, 27%, 0.10)', text: '#4D7C0F', border: 'hsla(80, 79%, 27%, 0.35)' },
  '其他': { bg: 'hsl(var(--surface-2))', text: 'hsl(var(--muted))', border: 'hsl(var(--border))' },
}

export function getFactorTypeColor(t) {
  return FACTOR_TYPE_COLORS[t] || FACTOR_TYPE_COLORS['其他']
}

File: smcore/strategy/factor_types.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""策略 → 因子类型映射（报告/监控层重分类用，2026-09-15）。

用户要求：报告与监控不再「按 5 策略名」分组，改为「按因子类型」分组。
本模块是**唯一**的映射来源，报告层（report.py）与监控层（factor_ic_monitor.py）
都从这里取，避免两处各写一套导致漂移。

设计原则（与「别过度工程 / 底层不动」一致）：
- 只做展示层的**标签与归并**，不触碰任何策略选股 / 融合 / 权重逻辑；
- 每策略归一个**主因子类型**（1:1），保证归并桶清晰、无歧义；
- 一个 pick 可命中多个策略（来源策略="Boll/Momentum"），因此可映射到多个因子类型——
  `factor_types_of_source` 返回去重后的类型列表，供明细展示与贡献度统计。

2026-09-15 修订：原「题材·事件」把 Theme 与 CCTV 合并，归因发现 Theme 实为拖累、
CCTV 才是正贡献，合并掩盖了内部分化 → 拆成「题材」(Theme) 与「事件·舆情」(CCTV) 两类。
"""
from __future__ import annotations

# 主因子类型映射：key 用小写策略 id（与 adaptive_weights.ALL_STRATEGIES 对齐）
STRATEGY_FACTOR_TYPE = {
    "boll": "反转·均值回归",
    "momentum": "动量",
    "relativity": "相对强度·资金流",
    "theme": "题材",
    "cctv": "事件·舆情",
    # 2026-09-17：原单一「基本面·质量价值」拆成三个正交基本面类型，便于横向对比各维 edge。
    "quality": "基本面·质量",
    "value": "基本面·估值",
    "size": "基本面·规模",
}

# 展示顺序（与既有策略认知一致，便于阅读）
FACTOR_TYPE_ORDER = [
    "动量",
    "反转·均值回归",
    "相对强度·资金流",
    "题材",
    "事件·舆情",
    "基本面·质量",
    "基本面·估值",
    "基本面·规模",
    "其他",
]

_DEFAULT_TYPE = "其他"


def _normalize(name: str) -> str:
    """策略名归一化：去空白、转小写，兼容 'Boll' / 'boll' / 'Momentum' 等写法。"""
    return str(name).strip().lower()


def factor_type_of(strategy: str, default: str = _DEFAULT_TYPE) -> str:
    """单个策略 → 主因子类型。"""
    return STRATEGY_FACTOR_TYPE.get(_normalize(strategy), default)


def factor_types_of_source(source: str, default: str = _DEFAULT_TYPE) -> list[str]:
    """来源策略串（如 'Boll/Momentum' 或 'cctv'）→ 去重后的因子类型列表。

    用于明细展示与贡献度统计：一个 pick 可能命中多个策略，故可能返回多个类型。
    """
    if not source:
        return [default]
    out: list[str] = []
    for part in str(source).replace("、", "/").split("/"):
        part = part.strip()
        if not part:
            continue
        t = factor_type_of(part, default)
        if t not in out:
            out.append(t)
    return out or [default]


def rollup_counts(counts_by_strategy: dict[str, int]) -> dict[str, int]:
    """{策略: 计数} → {因子类型: 计数} 归并（同类型策略计数相加）。

    归并顺序按 FACTOR_TYPE_ORDER，未列出的类型（理论不该出现）追加在末尾。
    """
    agg: dict[str, int] = {}
    for strategy, cnt in counts_by_strategy.items():
        t = factor_type_of(strategy)
        agg[t] = agg.get(t, 0) + (cnt or 0)
    ordered = {k: agg[k] for k in FACTOR_TYPE_ORDER if k in agg}
    for k in agg:  # 兜底：不在预定义顺序里的类型
        if k not in ordered:
            ordered[k] = agg[k]
    return ordered
